src/server/server_world.rs

```diff
@@ -0,0 +1,438 @@
+    type Item<'world, 'state> = ServerWorld<'world, 'state>;
+
+    fn init_state(world: &mut World, system_meta: &mut SystemMeta) -> Self::State {
+        let mut filtered_access = FilteredAccess::default();
+
+        let marker_id = world.register_component::<Replicated>();
+        filtered_access.add_component_read(marker_id);
+
+        let rules = world.resource::<ReplicationRules>();
+        debug!("initializing with {} replication rules", rules.len());
+        let combined_access = system_meta.component_access_set().combined_access();
+        for rule in rules.iter() {
+            for &(component_id, _) in &rule.components {
+                filtered_access.add_component_read(component_id);
+                assert!(
+                    !combined_access.has_component_write(component_id),
+                    "replicated component `{}` in system `{}` shouldn't be in conflict with other system parameters",
+                    world.components().get_name(component_id).unwrap(),
+                    system_meta.name(),
+                );
+            }
+        }
+
+        let access = filtered_access.access().clone();
+
+        // SAFETY: used only to extend access.
+        unsafe {
+            system_meta.component_access_set_mut().add(filtered_access);
+        }
+
+        ReplicationReadState {
+            access,
+            marker_id,
+            archetypes: Default::default(),
+            // Needs to be cloned because `new_archetype` only accepts the state.
+            rules: world.resource::<ReplicationRules>().clone(),
+        }
+    }
+
+    unsafe fn new_archetype(
+        state: &mut Self::State,
+        archetype: &Archetype,
+        system_meta: &mut SystemMeta,
+    ) {
+        if !archetype.contains(state.marker_id) {
+            return;
+        }
+
+        trace!("marking `{:?}` as replicated", archetype.id());
+        let mut replicated_archetype = ReplicatedArchetype::new(archetype.id());
+        for rule in state.rules.iter().filter(|rule| rule.matches(archetype)) {
+            for &(component_id, fns_id) in &rule.components {
+                // Since rules are sorted by priority,
+                // we are inserting only new components that aren't present.
+                if replicated_archetype
+                    .components
+                    .iter()
+                    .any(|component| component.component_id == component_id)
+                {
+                    continue;
+                }
+
+                let storage_type = archetype.get_storage_type(component_id).unwrap_unchecked();
+                replicated_archetype.components.push(ReplicatedComponent {
+                    component_id,
+                    storage_type,
+                    fns_id,
+                });
+            }
+        }
+
+        // Update system access for proper parallelization.
+        for component in &replicated_archetype.components {
+            let archetype_id = archetype
```

Comment:
Isn't this redundant? I don't see how this could add anything not added in `init_state`.